### ISSUE TITLE
Problem with generated SDF bitmap

### ISSUE THREAD

**Issue by BreyerW** [StbSharp/StbTrueTypeSharp#1]
Im trying to integrate your lib into my app and overall it works but SDF for certain characters is reeeeaaaly iffy. I tested several fonts: Arial, times, tahoma all of em have a bit of broken characters in resulting bitmap. Some characters looks like noise, some skewed, rest are ok

wrong results:

skewed n character

![n_times](https://user-images.githubusercontent.com/12634170/71449740-ad363080-2753-11ea-8abc-4621019cd28c.png)

character d looks like noise

![d_times](https://user-images.githubusercontent.com/12634170/71449751-f4242600-2753-11ea-87c5-8f9cf9fb9ec4.png)

good results just to show that sometimes it works as expected

A character (a little strange it selected a small though)

![A_times](https://user-images.githubusercontent.com/12634170/71449757-32b9e080-2754-11ea-8bc3-3c69985dca98.png)

S character

![s_times](https://user-images.githubusercontent.com/12634170/71449760-4402ed00-2754-11ea-8765-0915eba183f0.png)

iffines in files proves there is a problem with generating SDF itself rather than rendering since bytes are saved to disk basically as-is

this is how i generate SDF: (c is of type char and passed as-is)

```csharp
var index=StbTrueType.stbtt_FindGlyphIndex(face, c);
var bytes=StbTrueType.stbtt_GetGlyphSDF(face,StbTrueType.stbtt_ScaleForPixelHeight(face,24),index,1,180,180,&width,&height,&xoff,&yoff);
					StbTrueType.stbtt_GetCodepointHMetrics(face,c,&advance,&bearingX);
					var bitmap = new Span<byte>(bytes, width * height).ToArray();
					var invalidchars = Path.GetInvalidFileNameChars();
					bool allowed = true;
					foreach(var invalid in invalidchars)
					{
						allowed = invalid != c;
					}
					if (c!=':' && allowed)
						using (System.Drawing.Bitmap bmp = new System.Drawing.Bitmap(width, height, System.Drawing.Imaging.PixelFormat.Format8bppIndexed))
						{

							var bmpdata = bmp.LockBits(new System.Drawing.Rectangle(0, 0, width, height), System.Drawing.Imaging.ImageLockMode.ReadWrite, bmp.PixelFormat);
							System.Runtime.InteropServices.Marshal.Copy(bitmap, 0, bmpdata.Scan0, bitmap.Length);
							bmp.UnlockBits(bmpdata);
							bmp.Save(@"B:\Sharp.Engine3\Sharp.Core\bin\Debug\netcoreapp3.1\" + c+"_"+fontName + ".png");
						}
```
I tried several parameters ones you see there and ones that can be seen in stb_truetype original work, some of my own invention none of them produces fully correct bitmaps

Since i mentioned that character 'A' turned into 'a' in picture does it mean i must preprocess standard .NET char c into codepoint understandable for stb_truetype?

Other than that im completely lost. Any ideas what might be wrong?

EDIT

normalizing string and/or converting to codepoints like this didnt work:

```csharp
public static int[] ToCodePoints(string str)
{
    if (str == null)
        throw new ArgumentNullException("str");

    var codePoints = new List<int>(str.Length);
    for (int i = 0; i < str.Length; i++)
    {
        codePoints.Add(Char.ConvertToUtf32(str, i));
        if (Char.IsHighSurrogate(str[i]))
            i += 1;
    }

    return codePoints.ToArray();
}
```

**Comment by rds1983** [comment]
Sorry, I am no expert in neither text rendering or stb_truetype. Actually even term 'SDF' is new to me.
Maybe it makes sense to ask this question at original stb repo: https://github.com/nothings/stb

**Comment by BreyerW** [comment]
Then lets do it from different angle: I assume you used bitmap variants? (GetCodepointBitmap and the like) because they also produce incorrect results although not as glaring and i wonder if you had to translate char somehow before passing to stb

**Comment by rds1983** [comment]
Well, I maintain following project that uses StbTrueTypeSharp: https://github.com/rds1983/SpriteFontPlus

Specifically it uses stb_truetype here:
https://github.com/rds1983/SpriteFontPlus/blob/master/src/SpriteFontPlus/TtfFontBaker.cs
(static font with predefined characters set)

and here:
https://github.com/rds1983/SpriteFontPlus/blob/master/src/SpriteFontPlus/FontStashSharp/FontSystem.cs
(dynamic font)

So far it works fine.

**Comment by TechPizzaDev** [comment]
The culprit is presumably the copying.
> System.Runtime.InteropServices.Marshal.Copy(bitmap, 0, bmpdata.Scan0, bitmap.Length);

The GDI+ bitmap has a stride that is not always equal to the width.

```csharp
for (int y = 0; y < height; y++)
    Marshal.Copy(
        source: bitmap, 
        startIndex: y * width, 
        destination: bmpdata.Scan0 + y * bmpdata.Stride,
        length: width);
```
This should work better as it is aware of the bitmap stride.

**Comment by BreyerW** [comment]
@TechnologicalPizza unlikely as same or similar broken results was seen simply rendering bytes via shader bypassing gdi completely. Gdi was used purely to save to disk

**Comment by TechPizzaDev** [comment]
Maybe you could upload the fonts that you are experiencing problems with so others can test?
I've used SDF with a couple fonts and never got incorrect data.

**Comment by BreyerW** [comment]
@TechnologicalPizza I used built-in fonts like arial tahoma and new roman times. Nothing fancy really. Strange part is that most characters looked ok (you can see some sdf images on most top post) and only few showed skewiness, noise or the like

**Comment by TechPizzaDev** [comment]
Times New Roman: ![test d](https://user-images.githubusercontent.com/23627133/78265980-736c8700-7505-11ea-9a6b-21c5acc85ea1.png) ![test n](https://user-images.githubusercontent.com/23627133/78265984-749db400-7505-11ea-845a-e44917482b2b.png)

Arial:  ![test d](https://user-images.githubusercontent.com/23627133/78266200-bdee0380-7505-11ea-8e89-fb32de020976.png) ![test n](https://user-images.githubusercontent.com/23627133/78266198-bd556d00-7505-11ea-9178-c259f40b5e32.png)

Tahoma: ![test d](https://user-images.githubusercontent.com/23627133/78266363-fab9fa80-7505-11ea-85ce-8cfdba5793dc.png) ![test n](https://user-images.githubusercontent.com/23627133/78266365-fb529100-7505-11ea-9e98-fcbb82ed1d49.png)

Looks pretty nice on my end (that was 64px height, 24px works as well)
I used practically the same code as you, so I'm lost on what your problem can be.
Maybe my repo is on a newer version or some other part of your code is silently failing.

**Comment by BreyerW** [comment]
@TechnologicalPizza  Yeah the only conclusion i can have something is iffy with my rendering pipeline or parameters

Btw what parameters did you use for getglyph and did you preprocess char to codepoint or passed as-is?

**Comment by TechPizzaDev** [comment]
Used the same parameters as your example code (1 for padding, ScaleForPixelHeight 24px for size, 180 for onedge and distscale). 
Also used `StbTrueType.stbtt_FindGlyphIndex(face, character);` where `character` was just a C# `char`.
